services/director-v2/src/simcore_service_director_v2/utils/distributed_identifier.py

```diff
@@ -0,0 +1,281 @@
+            self._serialize_cleanup_context(cleanup_context),
+        )
+        return identifier, result
+
+    async def remove(self, identifier: Identifier, *, reraise: bool = False) -> None:
+        """Attempts to remove the resource, if an error occurs it is logged.
+
+        Arguments:
+            identifier -- user chosen identifier for the resource
+            reraise -- when True raises any exception raised by ``destroy`` (default: {False})
+        """
+
+        cleanup_context = await self._get_identifier_context(identifier)
+        if cleanup_context is None:
+            _logger.warning(
+                "Something went wrong, did not find any context for %s", identifier
+            )
+            return
+
+        with log_context(
+            _logger, logging.DEBUG, f"{self.__class__}: removing {identifier}"
+        ), log_catch(_logger, reraise=reraise):
+            await self._destroy(identifier, cleanup_context)
+
+        await self.redis_client_sdk.redis.delete(self._to_redis_key(identifier))
+
+    @classmethod
+    @abstractmethod
+    def _deserialize_identifier(cls, raw: str) -> Identifier:
+        """User provided deserialization for the identifier
+
+        Arguments:
+            raw -- stream to be deserialized
+
+        Returns:
+            an identifier object
+        """
+
+    @classmethod
+    @abstractmethod
+    def _serialize_identifier(cls, identifier: Identifier) -> str:
+        """User provided serialization for the identifier
+
+        Arguments:
+            cleanup_context -- user defined identifier object
+
+        Returns:
+            object encoded as string
+        """
+
+    @classmethod
+    @abstractmethod
+    def _deserialize_cleanup_context(cls, raw: str) -> CleanupContext:
+        """User provided deserialization for the context
+
+        Arguments:
+            raw -- stream to be deserialized
+
+        Returns:
+            an object of the type chosen by the user
+        """
+
+    @classmethod
+    @abstractmethod
+    def _serialize_cleanup_context(cls, cleanup_context: CleanupContext) -> str:
+        """User provided serialization for the context
+
+        Arguments:
+            cleanup_context -- user defined cleanup context object
+
+        Returns:
+            object encoded as string
+        """
+
+    @abstractmethod
+    async def is_used(
+        self, identifier: Identifier, cleanup_context: CleanupContext
+    ) -> bool:
+        """Check if the resource associated to the ``identifier`` is
+        still being used.
+        # NOTE: a resource can be created but not in use.
+
+        Arguments:
+            identifier -- user chosen identifier for the resource
+            cleanup_context -- user defined CleanupContext object
+
+        Returns:
+            True if ``identifier`` is still being used
+        """
+
+    @abstractmethod
+    async def _create(self, **extra_kwargs) -> tuple[Identifier, ResourceObject]:
```

Comment:
so many of these functions are abstract but protected. Kind o this pattern of non-virtual interfaces that was so popular in C++ for a while: instead of overriding the interface, the sub-class would override some protected members that would implement parts of the base class.


Nonetheless I find very confusing having `create` and `_create` in the same class...

Reply:
I had some issue with the naming. I'll explain my idea here. I want the user to implement generic method cor `creation`, `recovery` and `removal` of the resource.

Some of these methods need sone `Redis` calls. The user has to overwrite the methods , but doing so would also overwrite any existing context managers. 
So this was my middle ground solution.

The only thing I can do is have better naming.